[FE-17705] Dynamically update categorical legend domain

cmatzenbach opened the pull request:
### :speech_balloon: Description

This adds dynamic filtering to the legend for charts using a categorical legend.  Previously the original domain and range were being passed around in this part of the code, and no attempt at retrieving an updated domain was done for legends with categorical data.  Thus the legend displayed the initial domain, regardless if bbox/crossfilters/chart filters/etc were applied.  Now, when we detect categorical data, we make a call to crossfilter for the categorical field, get the updated topN values, pull the original range value (color) applied to each of the new topN values, and provide the legend with an updated domain and range.  This fixed the issue for pointmap, but did not for choropleth, due to a crossfilter bug (feature?).  That is fixed in this accompanying immerse PR: https://github.com/heavyai/immerse/pull/9240

### :page_facing_up: Jira Issue

Closes [FE-17705](https://heavyai.atlassian.net/browse/FE-17705)

### :mag: Verification Steps

1. Create a pointmap with a categorical column as the color measure
2. Zoom around the map, and make sure the legend updates with the values per the current bbox
3. Apply crossfilter, chart filters, dashboard filters, etc, making sure that the legend updates per the applied filters

### :camera_flash: Screenshot
Dynamic legend updating with standard range (no extra colors assigned by user):

https://github.com/heavyai/heavyai-charting/assets/15268289/a7fabf4f-632f-436a-9088-c9a7b07f6aa8

Now with custom range (extra color assignments added by user):

https://github.com/heavyai/heavyai-charting/assets/15268289/a526ba4c-47ae-4356-81ed-12cb531a4417


[FE-17705]: https://heavyai.atlassian.net/browse/FE-17705?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

cmatzenbach commented on src/chart-addons/stacked-legend.js:
```diff
@@ -119,83 +166,99 @@ export function getLegendStateFromChart(chart, useMap, selectedLayer) {
-            color_legend_descriptor = {
-              ...color,
-              domain: layer.colorDomain()
+            } else if (color.type === "ordinal") {
```
Ugly diff, due to the now-necessary `Promise.all` async wrapper.  Just did the async handling and added the condition below:

mapd-bot-os commented:
build failed

mapd-bot-os commented:
build failed

mapd-bot-os commented:
build failed

mapd-bot-os commented:
build failed

mapd-bot-os commented:
build failed

johallar commented on src/chart-addons/stacked-legend.js:
```diff
@@ -119,77 +156,93 @@ export function getLegendStateFromChart(chart, useMap, selectedLayer) {
+                    color.range,
+                    color.defaultOtherRange
+                  )
+                : undefined
```
This will try to destructure `undefined` and bomb if it hits the `else` condition
```suggestion
                : {}
```

mapd-bot-os commented:
build failed

johallar commented on src/chart-addons/stacked-legend.js:
```diff
@@ -98,7 +98,44 @@ function setColorScaleDomain_v1(domain) {
+        }
+      })
+      .catch(error => {
+        return null
```
I wonder if we want to let this error bubble up instead of swallowing? That's assuming it will get caught in the immerse saga, and do the generic chart error.

mapd-bot-os commented:
build failed

mapd-bot-os commented:
build failed

mapd-bot-os commented:
build failed

johallar commented on src/chart-addons/stacked-legend.js:
```diff
@@ -98,7 +98,42 @@ function setColorScaleDomain_v1(domain) {
+      .group()
+      .topAsync(size, OFFSET)
+      .then(results => {
+        if (results) {
+          return results.map(result => result.key0)
+        } else {
+          return null
+        }
```
golfing: 
```suggestion
        return results?.map(result => result.key0) ?? null
```

johallar commented on src/chart-addons/stacked-legend.js:
```diff
@@ -98,7 +98,42 @@ function setColorScaleDomain_v1(domain) {
+      })
+      .catch(error => null)
+  } else {
+    return Promise.resolve(null)
```
I didn't think you had to explicitly return a promise here, since its an async function. Maybe charting babel config is weird tho

```suggestion
    return null
```

johallar commented:
Noticed this today while testing with multilayer choropleth charts. The legend will disappear when collapsing a layer's section. Tested on master on singularity and seems to work so I think this may be a regression: http://singularity.mapd.com:6273/omnisci/dashboard/527?tab=-Nvh3VToxos7YFaKnmW5

https://github.com/heavyai/heavyai-charting/assets/1031659/ec0dab20-03f6-485e-b127-bf3ba52caf2e

mapd-bot-os commented:
build failed

mapd-bot-os commented:
build failed

mapd-bot-os commented:
build failed

cmatzenbach commented:
Ok @johallar the crossfilter issue with the stacked legend is now fixed.  I had moved the code over to immerse, but that caused _way_ more issues downstream.  It is best handled in charting, as immerse is (for better or worse) designed around categorical domains _not_ being updated on map move/zoom, and doing so causes a ton of issues.  By keeping it in charting, the code is isolated and only impacts the legend, which was the customer issue.  Ready for re-review.

cmatzenbach commented:
Screencast showing crossfilter bug fixed:

https://github.com/heavyai/heavyai-charting/assets/15268289/e7ed95cd-8b10-4d83-ba42-078c0015725a

mapd-bot-os commented:
build failed

mapd-bot-os commented:
build failed

mapd-bot-os commented:
build failed

mapd-bot-os commented:
build failed
